src/Input/Select/Select.tsx

```diff
@@ -15,346 +15,271 @@ import {
-      const activeChild = filterValue[activeElementIndex];
-      const onOptionClick = get(activeChild, 'props.onOptionClick');
+  const getActiveElement = React.useCallback(() => {
+    return node.current.querySelector('.active') as HTMLLIElement;
+  }, []);
 
-      this.setState({
-        selectedValue: activeElement.innerText,
-        filterValue: React.Children.map(children, data => data),
-        floating: true,
-        isFocus: false,
-      });
-      if (onChange !== undefined) {
-        const itemValue = activeElement.dataset.value;
-        onChange(itemValue);
-      }
-      if (onOptionClick !== undefined) {
-        onOptionClick(e);
-      }
-    } else if (e.keyCode === 27) {
-      inputElement.blur();
-      this.setState({ isFocus: false });
-    }
-  };
+  // Should be called when any key is pressed inside the component
+  const handleKeyDown = React.useCallback(
+    (e: React.KeyboardEvent<HTMLInputElement>) => {
+      const inputElement = e.target as HTMLInputElement;
 
-  handleMouseEnter = (e: React.MouseEvent<HTMLLIElement, MouseEvent>) => {
-    const listItemElement = e.target as HTMLLIElement;
-    this.setState({
-      cursor: Number(listItemElement.dataset.id),
-    });
-  };
+      if (disableTyping) {
+        e.preventDefault();
+      }
 
-  scrollToActiveElement = () => {
-    const selectListElement = this.node.current.querySelector(
-      '.select-listbox'
-    ) as HTMLUListElement;
-    selectListElement.scrollTop = this.getActiveElement().offsetTop;
-  };
-
-  getActiveElement = () => {
-    return this.node.current.querySelector('.active') as HTMLLIElement;
-  };
-
-  render() {
-    const {
-      label,
-      status,
-      disabled,
-      className,
-      onFocus,
-      onBlur,
-      onChange,
-      noOptionResult = 'No results found',
-      small,
-      disableTyping,
-      removeFloatingLabel,
-      removeDropIcon,
-      error,
-      renderError,
-      value, // eslint-disable-line @typescript-eslint/no-unused-vars
-      children, // eslint-disable-line @typescript-eslint/no-unused-vars
-      ...defaultProps
-    } = this.props;
-
-    const {
-      floating,
-      isFocus,
-      selectedValue = '',
-      filterValue,
-      cursor,
-      isLoading,
-    } = this.state;
-
-    const deprecatedStatus = status || (error && 'error');
-
-    const shouldShowError = error && typeof error !== 'boolean';
-
-    let completeError = null;
-    if (shouldShowError) {
-      if (renderError) {
-        completeError = renderError(error);
-      } else {
-        completeError = (
-          <SelectErrorDefault className="select-error-default">
-            {error}
-          </SelectErrorDefault>
-        );
+      if (e.keyCode === 38 && cursor > 0) {
+        setCursor(cursor - 1);
+        scrollToActiveElement();
+      } else if (e.keyCode === 40 && cursor < filterValue.length - 1) {
+        setCursor(cursor + 1);
+        scrollToActiveElement();
```

Comment:
Saw the original usage is to call `scrollToActiveElement` after `cursor` was updated.

But the current one seems to call them concurrently.

Reply:
Good catch, I wasn't familiar with the second `setState` argument and didn't realise that this should be sequential. I added an `effect` that triggers the scroll _after_ `cursor` updates.